logstash-core/src/main/java/org/logstash/instrument/metrics/ExtendedFlowMetric.java

```diff
@@ -0,0 +1,360 @@
+    int estimateCapturesRetained() {
+        return this.retentionWindows.orElse(Collections.emptyList())
+                                    .stream()
+                                    .map(RetentionWindow::estimateSize)
+                                    .mapToInt(Integer::intValue)
+                                    .sum();
+    }
+
+    /**
+     * Internal introspection for tests to measure excess retention.
+     * @param retentionWindowFunction given a policy, return a retention window, in nanos
+     * @return the sum of over-retention durations.
+     */
+    Duration estimateExcessRetained(final ToLongFunction<FlowMetricRetentionPolicy> retentionWindowFunction) {
+        final long currentNanoTime = nanoTimeSupplier.getAsLong();
+        final long cumulativeExcessRetained =
+                this.retentionWindows.orElse(Collections.emptyList())
+                                     .stream()
+                                     .map(s -> s.excessRetained(currentNanoTime, retentionWindowFunction))
+                                     .mapToLong(Long::longValue)
+                                     .sum();
+        return Duration.ofNanos(cumulativeExcessRetained);
+    }
+
+    /**
+     * A {@link RetentionWindow} efficiently holds sufficient {@link FlowCapture}s to
+     * meet its {@link FlowMetricRetentionPolicy}, providing access to the youngest capture
+     * that is older than the policy's allowed retention (if any).
+     * The implementation is similar to a singly-linked list whose youngest captures are at
+     * the tail and oldest captures are at the head, with an additional pre-tail stag.
+     * Compaction is always done at read-time and occasionally at write-time.
+     * Both reads and writes are concurrency-safe.
+     */
+    private static class RetentionWindow {
+        private final AtomicReference<FlowCapture> stagedCapture = new AtomicReference<>();
+        private final AtomicReference<Node> tail;
+        private final AtomicReference<Node> head;
+        private final FlowMetricRetentionPolicy policy;
+
+        RetentionWindow(final FlowMetricRetentionPolicy policy, final FlowCapture zeroCapture) {
+            this.policy = policy;
+            final Node zeroNode = new Node(zeroCapture);
+            this.head = new AtomicReference<>(zeroNode);
+            this.tail = new AtomicReference<>(zeroNode);
+        }
+
+        /**
+         * Append the newest {@link FlowCapture} into this {@link RetentionWindow},
+         * while respecting our {@link FlowMetricRetentionPolicy}.
+         * We tolerate minor jitter in the provided {@link FlowCapture#nanoTime()}, but
+         * expect callers of this method to minimize lag between instantiating the capture
+         * and appending it.
+         *
+         * @param newestCapture the newest capture to stage
+         */
+        private void append(final FlowCapture newestCapture) {
+            final Node casTail = this.tail.get(); // for CAS
```

Comment:
NIT: we only need `AtomicReference<Node>#getAcquire()` here, not a full volatile read.

```suggestion
            final Node casTail = this.tail.getAcquire(); // for CAS
```